File: stack-bench/backends/ruby-rails/app/controllers/bench_controller.rb
class BenchController < ApplicationController
  include ActionController::Live

  REGIONS = %w[us-east us-west eu-central].freeze
  STATUSES = %w[ok warn err].freeze

  def rows
    render json: (0...200).map { |i| make_row(i) }
  end

  def act
    render json: { ok: true, ts: (Time.now.to_f * 1000).to_i }
  end

  def events
    response.headers["Content-Type"] = "text/event-stream"
    response.headers["Cache-Control"] = "no-cache"
    i = 0
    loop do
      response.stream.write("data: #{ { seq: i, ts: (Time.now.to_f * 1000).to_i }.to_json }\n\n")
      i += 1
      sleep 0.1
    end
  rescue IOError, ActionController::Live::ClientDisconnected
  ensure
    response.stream.close
  end

  private

  def make_row(i)
    {
      id: i,
      service: "svc-#{i % 12}",
      region: REGIONS[i % 3],
      status: STATUSES[i % 3],
      latency_ms: 20 + ((i * 37) % 400),
      rps: 100 + ((i * 91) % 5000),
      updated_at: Time.at(1_752_900_000 + i).utc.iso8601
    }
  end
end

File: stack-bench/backends/elixir-phoenix/lib/bench_phx_web/controllers/bench_controller.ex
defmodule BenchPhxWeb.BenchController do
  use BenchPhxWeb, :controller

  @regions ~w(us-east us-west eu-central)
  @statuses ~w(ok warn err)

  defp make_row(i) do
    %{
      id: i,
      service: "svc-#{rem(i, 12)}",
      region: Enum.at(@regions, rem(i, 3)),
      status: Enum.at(@statuses, rem(i, 3)),
      latency_ms: 20 + rem(i * 37, 400),
      rps: 100 + rem(i * 91, 5000),
      updated_at: DateTime.from_unix!(1_752_900_000 + i) |> DateTime.to_iso8601()
    }
  end

  def rows(conn, _params) do
    json(conn, Enum.map(0..199, &make_row/1))
  end

  def act(conn, _params) do
    json(conn, %{ok: true, ts: System.system_time(:millisecond)})
  end

  def events(conn, _params) do
    conn =
      conn
      |> put_resp_content_type("text/event-stream")
      |> put_resp_header("cache-control", "no-cache")
      |> send_chunked(200)

    stream_events(conn, 0)
  end

  defp stream_events(conn, i) do
    payload = Jason.encode!(%{seq: i, ts: System.system_time(:millisecond)})

    case chunk(conn, "data: #{payload}\n\n") do
      {:ok, conn} ->
        Process.sleep(100)
        stream_events(conn, i + 1)

      {:error, _} ->
        conn
    end
  end
end
